### ISSUE TITLE
Re-execute based on given results

### ISSUE THREAD

**Issue by The-Mule** [teemtee/tmt#2332]
Tests are flaky, especially when executed in batches. At some point re-execution of tests is needed to avoid false positives. Currently there is no easy way to do so. One approach is to run the same plan with specific test selection. However from the users perspective this is far from perfect. It would be nice to have a way to tell TMT to rerun a plan based on its own results (ideally allowing all supported report methods). I can see many traps along the way. We need to make sure there is a way to keep specific tests always in the plan ("init-tasks" and "fini-tasks") somehow, for instance.

**Comment by kkaarreell** [comment]
I find this a design flaw in tmt/TF infrastructure. It is due to a fact that tests are fetch from a repository and identified dynamically based on fmf metadata. And this is done by tmt running in TF. That means that if there are supposed to be any changes in the plan it must be reflected in the repository itself. We would basically have to create a new plan that will run the modified set of tests.
This is something `tmt` wasn't built to do. In other words, we do not have tools for "dynamic" test plan creation and adjustments (e.g. based on former test results). Another problematic point might be identifying tests to skip and reschedule, given the fact that each of them may be in the plan multiple times and only some test iterations would be (un)desired.
Comparing this to Beaker and `wow`, here the recipe is generated on a user side and submitted to Beaker. Therefore it is way easier to incorporate modifications into the recipe. Unlike `tmt` which reads all these things dynamically from a test repo.

One thing that would be able to bring this closer to how Beaker works would be the ability to export tmt plans to a file and pass TF such a tmt plan(s) from the outside (e.g. by uploading the file), not having to commit stuff info some repository just to make it available to TF. This would leave a space for some tooling that would do the plan pre-processing (e.g. taking former results into account when generating a new plan). Yes, one could do that today and commit such a plan to some repo, but I am not sure if this would be a step in the right direction.

**Comment by kkaarreell** [comment]
Btw, wouldn't it be nice if we would be able to map test results (results.yaml) to tmt metadata of individual tests so that one could generate and filter plans utilizing those test results? Just imagine you would have a `discover` step which would filter tests according to former test results. When rescheduling tests, I could tell TF/tmt to apply an additional filter to individual discover steps to preserve "setup" tasks (e.g. based on a tag) and preserve failed tests (`results != pass`).

**Comment by happz** [comment]
> Btw, wouldn't it be nice if we would be able to map test results (results.yaml) to tmt metadata of individual tests so that one could generate and filter plans utilizing those test results? Just imagine you would have a `discover` step which would filter tests according to former test results. When rescheduling tests, I could tell TF/tmt to apply an additional filter to individual discover steps to preserve "setup" tasks (e.g. based on a tag) and preserve failed tests (`results != pass`).

What would you need to achieve that? There are `results.yaml` and `tests.yaml`, both per plan, containing results and corresponding test metadata. Test name and serial numbers should be enough to match the result and its test (or test and its results). So, this looks like good enough info to help with this particular feature, what would be needed besides that?

**Comment by kkaarreell** [comment]
> > Btw, wouldn't it be nice if we would be able to map test results (results.yaml) to tmt metadata of individual tests so that one could generate and filter plans utilizing those test results? Just imagine you would have a `discover` step which would filter tests according to former test results. When rescheduling tests, I could tell TF/tmt to apply an additional filter to individual discover steps to preserve "setup" tasks (e.g. based on a tag) and preserve failed tests (`results != pass`).
> 
> What would you need to achieve that? There are `results.yaml` and `tests.yaml`, both per plan, containing results and corresponding test metadata. Test name and serial numbers should be enough to match the result and its test (or test and its results). So, this looks like good enough info to help with this particular feature, what would be needed besides that?

I think the answer depends on at what level you would like to implement it. In `tmt` itself? Or some other tool that will generate `tmt` plans? For the former, I would need `tmt` to be able to read results.yaml and update respective nodes in the `fmf` tree with addional attributes representing the former test result. While results.yaml contain some details, it alone doesn't contain enough data to do a rerun. I don't know how difficult it would be to make such a "merge" and end up with a "plan" that can be scheduled by tmt again.

**Comment by psss** [comment]
Planning meeting notes:

 * Extend the `discover` step to consult the `report-portal` plugin for an additional filter for test discovery.
 * Clarify what exactly should happen when the new run finishes (e.g. should the original test result be overridden?)

**Comment by thrix** [comment]
For this to work in Testing Farm it would be sufficient if the launch ID to be restarted would need to be passed during restart, ideally via some tmt environment variables for now.

There is some concern how to map the test results from the launch to the executed plans, and if it makes even sense.

**Comment by thrix** [comment]
Talk to @brakogh, @kkaarreell  and others and find a resonable way forward for this feature in 1.52, so it can get implemented later.

**Comment by kkaarreell** [comment]
Just dropping a few more thoughts on this.

1. I don not see a reason why RP should be involved. That would in fact make everything more fragile (reverse RP suite -> tmt plan mapping) and there is no reason to make RP a required piece in the overall architecture. Should be tmt/TF solution only.
2. It should be clarified if we want to re-run individual TF requests (the primary user ask) or only one particular `tmt` test plan. The former would make the implementation slightly more difficult but at the same time way more useful.
3. I still find `tmt` being able to schedule tests and plans based on a recipe-like input a crucial piece in this puzzle. Such a recipe could be something combining `tests.yaml` and `results.yaml`, possible extending it with additional metadata. But it would be a single source of data (sort of partially replacing `discover`) that TF and `tmt` would use for request/test scheduling.
4. TF would need to support this somehow either way. It is like restarting a TF request, some extra parameters from CLI or credentials needs to be handled by TF.

**Comment by psss** [comment]
This will take more time and is blocked on the almighty install. In the next sprint we will outline the tmt recipe specification #3854 as the first step towards the solution.
